src/sources/aws_s3/sqs.rs

```diff
@@ -0,0 +1,733 @@
+
+    async fn handle_sqs_message(
+        &self,
+        message: Message,
+        out: Pipeline,
+    ) -> Result<(), ProcessingError> {
+        let s3_event: S3Event = serde_json::from_str(message.body.unwrap_or_default().as_ref())
+            .context(InvalidSqsMessage {
+                message_id: message.message_id.unwrap_or("<empty>".to_owned()),
+            })?;
+
+        self.handle_s3_event(s3_event, out).await
+    }
+
+    async fn handle_s3_event(
+        &self,
+        s3_event: S3Event,
+        out: Pipeline,
+    ) -> Result<(), ProcessingError> {
+        for record in s3_event.records {
+            self.handle_s3_event_record(record, out.clone()).await?
+        }
+        Ok(())
+    }
+
+    async fn handle_s3_event_record(
+        &self,
+        s3_event: S3EventRecord,
+        out: Pipeline,
+    ) -> Result<(), ProcessingError> {
+        if s3_event.event_name.kind != "ObjectCreated" {
+            emit!(SqsS3EventRecordIgnoredInvalidEvent {
+                bucket: &s3_event.s3.bucket.name,
+                key: &s3_event.s3.object.key,
+                kind: &s3_event.event_name.kind,
+                name: &s3_event.event_name.name,
+            });
+            return Ok(());
+        }
+
+        // S3 has to send notifications to a queue in the same region so I don't think this will
+        // actually ever be it unless messages are being forwarded from one queue to another
+        if self.region.name() != s3_event.aws_region {
+            return Err(ProcessingError::WrongRegion {
+                bucket: s3_event.s3.bucket.name.clone(),
+                key: s3_event.s3.object.key.clone(),
+                region: s3_event.aws_region,
+            });
+        }
+
+        let object = self
+            .s3_client
+            .get_object(GetObjectRequest {
+                bucket: s3_event.s3.bucket.name.clone(),
+                key: s3_event.s3.object.key.clone(),
+                ..Default::default()
+            })
+            .await
+            .context(GetObject {
+                bucket: s3_event.s3.bucket.name.clone(),
+                key: s3_event.s3.object.key.clone(),
+            })?;
+
+        let metadata = object.metadata;
+        let timestamp = object
+            .last_modified
+            .and_then(|t| {
+                DateTime::parse_from_rfc2822(&t)
+                    .map(|ts| Utc.timestamp(ts.timestamp(), ts.timestamp_subsec_nanos()))
+                    .ok()
+            })
+            .unwrap_or_else(Utc::now);
+
+        match object.body {
+            Some(body) => {
+                let r = super::s3_object_decoder(
+                    self.compression,
+                    &s3_event.s3.object.key,
+                    object.content_encoding.as_deref(),
+                    object.content_type.as_deref(),
+                    body,
+                );
+
+                // Record the read error saw to propagate up later so we avoid ack'ing the SQS
```

Comment:
Ideas here also welcome :smile:

Reply:
I'm not really sure what the issue is here?

Reply:
It just feels a little messy to me to record intermediate state rather than just streaming. We ended up having a conversation about it in discord though: https://discordapp.com/channels/742820443487993987/751107681217019944/771486734110162964

My conclusion from that conversation is that it'd be nice to use more `TryStream`s and have `send_all` stop at the first error and return it so that I could flow all errors and then handle them in one place.